src/avro/writer.rs

```diff
@@ -100,7 +100,30 @@ impl<W: Write> Writer<W> {
         } else {
             0
         };
+        self.append_inner(n, value)
+    }
+
+    /// Append, but useful for creating files that are less-robust to improve round-trip testing
+    ///
+    /// This is public but only for internal use, its API might change without warning.
+    #[doc(hidden)]
+    pub fn __append_for_testing<T: ToAvro>(
+        &mut self,
+        value: T,
+        include_codec: bool,
+    ) -> Result<usize, Error> {
```

Comment:
This API makes me uncomfortable. It is undocumented but public, has `bool` parameters (which make call sites hard to read), and is easy to misuse: if you call `__append_for_testing` twice, the second call will have no effect. It also adds more complexity to the writer, just to expose this testing knob, than was required for the fix in the reader in the first place.

I think there's a pretty natural place to introduce this additional bit of state, which is the codec. If you make the writer store an `Option<Codec>` instead of a `Codec` directly, and have `with_codec` take an `Option<Codec>`, then `None` can indicate that you wish to omit the `avro.codec` metadata field entirely. You'll only need one additional conditional, in `flush`, and none of the other code will need to change.

Reply:
I agree that this API is ugly, and I originally did write this the way that you suggest.

But if we change the writer API to make it more elegant/orthogonal we make it easier to mess up and not include the codec in production use cases, which will silently do whatever you told it to do, which is not compress. There isn't any production use case that I can think of that actively wants that, and until we do have a reason to I'd rather keep the more public API safer and make "internal" APIs uglier.

I agree about the robustness issue, I'll improve the comment explaining the header behavior.